velox/functions/sparksql/ToPrettyString.h

```diff
@@ -0,0 +1,187 @@
+        auto output = DATE()->toString(input);
+        result.append(output);
+        return;
+      }
+    }
+    const auto castResult =
+        util::Converter<TypeKind::VARCHAR, void, util::SparkCastPolicy>::
+            tryCast(input);
+    if (castResult.hasError()) {
+      result.setNoCopy(detail::kNull);
+    } else {
+      result.copy_from(castResult.value());
+    }
+  }
+
+  template <typename TInput>
+  void callNullable(out_type<Varchar>& result, const TInput* a) {
+    if (a) {
+      call(result, *a);
+    } else {
+      result.setNoCopy(detail::kNull);
+    }
+  }
+
+ private:
+  TypePtr inputType_;
+};
+
+template <typename TExec>
+struct ToPrettyStringTimeStampFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExec);
+
+  void initialize(
+      const std::vector<TypePtr>& /*inputTypes*/,
+      const core::QueryConfig& config,
+      const arg_type<Timestamp>* /*timestamp*/) {
+    auto timezone = config.sessionTimezone();
+    if (!timezone.empty()) {
+      options_.timeZone = date::locate_zone(timezone);
+    }
+    timestampRowSize_ = getMaxStringLength(options_);
+  }
+
+  void call(out_type<Varchar>& result, const Timestamp& input) {
+    Timestamp inputValue(input);
+    if (options_.timeZone) {
+      inputValue.toTimezone(*(options_.timeZone));
+    }
+    result.reserve(timestampRowSize_);
+    const auto stringView =
+        Timestamp::tsToStringView(inputValue, options_, result.data());
+    result.resize(stringView.size());
+  }
+
+  void callNullable(
+      out_type<Varchar>& result,
+      const arg_type<Timestamp>* timestamp) {
+    if (timestamp) {
+      call(result, *timestamp);
+    } else {
+      result.setNoCopy(detail::kNull);
+    }
+  }
+
+ private:
+  TimestampToStringOptions options_ = {
+      .precision = TimestampToStringOptions::Precision::kMicroseconds,
+      .leadingPositiveSign = true,
+      .skipTrailingZeros = true,
+      .zeroPaddingYear = true,
+      .dateTimeSeparator = ' ',
+  };
+  std::string::size_type timestampRowSize_;
+};
+
+template <typename TExec>
+struct ToPrettyStringDecimalFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExec);
+
+  template <typename A>
+  void initialize(
+      const std::vector<TypePtr>& inputTypes,
+      const core::QueryConfig& /*config*/,
+      A* /*a*/) {
+    const auto [precision, scale] = getDecimalPrecisionScale(*inputTypes[0]);
+    precision_ = precision;
+    scale_ = scale;
+    maxRowSize_ = precision + 1;
+    if (scale > 0) {
+      ++maxRowSize_; // A dot.
+    }
+    if (precision == scale) {
+      ++maxRowSize_; // Leading zero.
+    }
+  }
+
+  template <typename TInput>
+  void call(out_type<Varchar>& result, const TInput& input) {
+    result.reserve(maxRowSize_);
+    auto view = exec::detail::convertToStringView<TInput>(
+        input, scale_, maxRowSize_, result.data());
+    if (view.isInline()) {
+      result.setNoCopy(view);
+      result.resize(0);
+    } else {
+      result.resize(view.size());
```

Comment:
I think there is not need for if-else block. You can just do this unconditionally.

Reply:
This is for decimal, min long decimal value `-1'000'000'000'000'000'000 + 1` with `.` is not inline